src/main/java/org/opensearch/security/action/apitokens/ApiTokenAction.java

```diff
@@ -295,4 +354,99 @@ private void sendErrorResponse(RestChannel channel, RestStatus status, String er
+                clusterPermissionsWithoutActionGroups.removeIf(matcher);
+            }
+        }
+
+        if (!clusterPermissionsWithoutActionGroups.isEmpty()) {
+            throw new OpenSearchException("User does not have all requested cluster permissions");
+        }
+
+        // Verify user has all requested index permissions
+        for (ApiToken.IndexPermission requestedPermission : indexPermissions) {
+            // First, flatten/resolve any action groups into the underlying actions and remove action group names (which may not exact
+            // match)
+            Set<String> resolvedActions = new HashSet<>(flattenedActionGroups.resolve(requestedPermission.getAllowedActions()));
+            resolvedActions.removeIf(permission -> actionGroupsConfiguraiton.getCEntries().containsKey(permission));
+
+            // For each index pattern in the requested permission
+            for (String requestedPattern : requestedPermission.getIndexPatterns()) {
+
+                Set<String> actionsForIndexPattern = new HashSet<>(resolvedActions);
+
+                // Check each role the user has
+                for (String roleName : roles) {
+                    RoleV7 role = (RoleV7) rolesConfiguration.getCEntry(roleName);
+                    if (role == null || role.getIndex_permissions() == null) continue;
+
+                    // Check each index permission block in the role
+                    for (RoleV7.Index indexPermission : role.getIndex_permissions()) {
+                        List<String> rolePatterns = indexPermission.getIndex_patterns();
+                        List<String> roleIndexPerms = indexPermission.getAllowed_actions();
+
+                        // Check if this role's pattern covers the requested pattern
+                        if (WildcardMatcher.from(rolePatterns).test(requestedPattern)) {
+                            // Get resolved actions for this role's index permissions
+                            Set<String> roleActions = flattenedActionGroups.resolve(roleIndexPerms);
+                            WildcardMatcher matcher = WildcardMatcher.from(roleActions);
+
+                            actionsForIndexPattern.removeIf(matcher);
+                        }
+                    }
+                }
+
+                // After checking all roles, verify if all requested actions were covered
+                if (!actionsForIndexPattern.isEmpty()) {
+                    throw new OpenSearchException("User does not have sufficient permissions for index pattern: " + requestedPattern);
+                }
+            }
+        }
+    }
+
+    private SecurityDynamicConfiguration<?> load(final CType<?> config) {
+        SecurityDynamicConfiguration<?> loaded = configurationRepository.getConfiguration(config);
+        return DynamicConfigFactory.addStatics(loaded);
+    }
+
+    protected void authorizeSecurityAccess(RestRequest request) throws IOException {
```

Comment:
What do you think about having a dedicated transport action backing the REST action? Then, one would get access control out of the box without needing explicit checks within the action code.

Reply:
Agree with this though it depends on the action.

If the security admin is going to a page in the security dashboards plugin to manage the existing tokens (listing them and performing actions such as revoke) then these should follow the authorization model of other security APIs.

If regular users are allowed to request a token (with a subset of their own permissions) then that could be authorized like other cluster actions in OpenSearch.

Reply:
I think (for now) my operating assumption would be that this is an admin-level action to issue api tokens to connect with other services programmatically. Since API tokens (currently) do not follow the new optimized privileges evaluation, and thus have higher performance penalty than other means of authc/z, and thus we should restrict this as much as possible. However, if there is community feedback/requests for this feature, then we can consider it at a later time. Since we want to authorize these APIs like the others, I added this logic.

Reply:
To be honest, it is a bit difficult for me to review this without a proper scope definition. During review, I learned from you about a couple of assumptions and assumed limitations (such as not using the optimized privilege evaluation code paths, not supporting regexes, etc).  IMHO, there should be some definitions in the issue #1504 what's the initial goal to be achieved and what not. That will be then an information to guide the code review process. Otherwise, this review process will get quite inefficient. 

If the assumption is indeed that issuing tokens is an admin-level action, I do wonder why we need the matching to existing privileges at all. Won't an admin-level action just be similar to granting new privileges?

Reply:
If its rolled out to admin users initially at first than I agree, but if the longer term plan is for other users to be able to request tokens then the logic to determine whether the requested permissions are a subset of the user's permissions would be needed. I was thinking to keep the logic simple and only consider: 1) `cluster_permissions` and 2) `index_permissions`. For index permissions, would it be sufficient to limit the checks to concrete indices and glob-like patterns and prohibit `regexes`?

In practice, I have not come across many roles definitions where cluster administrators use regex when define the index patterns in a role. I'm sure its used, but not prevalent from my experience.